src/mem/mem_scheduler.cc

```diff
@@ -0,0 +1,402 @@
+/*
+ * Copyright (c) 2017 Jason Lowe-Power
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <string>
+#include "mem/mem_scheduler.hh"
+
+#include "base/trace.hh"
+#include "debug/MemScheduler.hh"
+
+MemScheduler::MemScheduler(const MemSchedulerParams &params):
+    ClockedObject(params),
+    nextReqEvent([this]{ processNextReqEvent(); }, name()),
+    nextRespEvent([this]{ processNextRespEvent(); }, name()),
+    readBufferSize(params.read_buffer_size),
+    writeBufferSize(params.write_buffer_size),
+    respBufferSize(params.resp_buffer_size),
+    nMemPorts(params.port_mem_side_connection_count),
+    nCpuPorts(params.port_cpu_side_connection_count)
+{
+
+    panic_if(readBufferSize == 0, "readBufferSize should be non-zero");
+    panic_if(writeBufferSize == 0, "writeBufferSize "
+                                    "should be non-zero");
+    for (uint32_t i = 0; i < nCpuPorts; ++i){
+        cpuPorts.emplace_back(name() + ".cpu_side" + std::to_string(i), i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize, i));
```

Comment:
Why push_back and not emplace? It's probably functionally the same, but push_back is creating a copy.